src/mame/drivers/magicard.cpp

```diff
@@ -118,54 +199,194 @@
+
+	DECLARE_WRITE_LINE_MEMBER(scc66470_irq);
+	DECLARE_WRITE_LINE_MEMBER(cpu_i2c_scl);
+	DECLARE_WRITE_LINE_MEMBER(cpu_i2c_sda_write);
+	DECLARE_READ_LINE_MEMBER(cpu_i2c_sda_read);
+	
+	void update_sda(uint8_t device, uint8_t state);
+	void update_scl(uint8_t device, uint8_t state);
+
+
+private:
+
+};
+
+class magicard_state : public magicard_base_state
+{
+public:
+	magicard_state(const machine_config &mconfig, device_type type, const char *tag)
+		: magicard_base_state(mconfig, type, tag)
+		, m_nvram(*this, "nvram")
+		, m_ds1207(*this, "ds1207")
+
+	{ }
+
+	void magicard_pic54(machine_config &config);
+	void magicard(machine_config &config);
+	void unkte06(machine_config &config);
+	
+protected:
+	virtual void machine_start() override;
+
+private:
+	required_device<nvram_device>    m_nvram;
+	required_device<ds1207_device> m_ds1207;
+	
+	void magicard_map(address_map &map);
+	uint8_t nvram_r(offs_t offset);
+	void nvram_w(offs_t offset, uint8_t data);
+	uint8_t read_ds1207(offs_t offset);
+	void write_ds1207(offs_t offset, uint8_t data);
+	
+	std::unique_ptr<uint8_t[]>   m_nvram8;
+};
+
+class hotslots_state : public magicard_base_state
+{
+public:
+	hotslots_state(const machine_config &mconfig, device_type type, const char *tag)
+		: magicard_base_state(mconfig, type, tag)
+		, m_ds2401(*this, "serial_id")
+		, m_ds1207(*this, "ds1207")
+		, m_rtc(*this, "rtc")
+
+
+	{ }
+
+	void hotslots_base(machine_config &config);
+	void hotslots(machine_config &config);
+	void hotslots_pic54(machine_config &config);
+	void magicle(machine_config &config);
+	void puzzleme(machine_config &config);
+
+protected:
+	virtual void machine_start() override;
+
+private:
+
+	optional_device<ds2401_device> m_ds2401;
+	optional_device<ds1207_device> m_ds1207;
+	required_device<rtc72421_device> m_rtc;
+
+	void hotslots_map_base(address_map &map);
+	void hotslots_map(address_map &map);
+	void puzzleme_map(address_map &map);
+	
+	uint8_t read_ds1207_ds2401(offs_t offset);
+	void write_ds1207_ds2401(offs_t offset, uint8_t data);
+	void output_w(offs_t offset, uint16_t data);
+
+	DECLARE_WRITE_LINE_MEMBER(cpu_int1);
 };
 
+#define NVRAM_SIZE ( 16384 )
+
+void magicard_base_state::machine_start()
+{
+	save_item(NAME(m_sda_state));
+	save_item(NAME(m_scl_state));
+}
+
+void magicard_state::machine_start()
+{
+	magicard_base_state::machine_start();
+	m_nvram8 = std::make_unique<uint8_t[]>(NVRAM_SIZE);
+	m_nvram->set_base(m_nvram8.get(),NVRAM_SIZE);
+}
+
+void hotslots_state::machine_start()
+{
+	 magicard_base_state::machine_start();
+}
 
-/*********************************************
-*               Video Hardware               *
-*********************************************/
+void magicard_state::nvram_w(offs_t offset, uint8_t data)
+{
+	if(offset < NVRAM_SIZE)
+	{
+		m_nvram8[ offset ] = data ;
+	}
+}
+
+uint8_t magicard_state::nvram_r(offs_t offset)
+{
+	if(offset < NVRAM_SIZE)
+	{
+		return m_nvram8[ offset ] ;
+	}
+	else
+	{
+		return 0 ;
+	}
```

Comment:
You should return the unmapped read value for the address space in unmapped areas (or just not map the handler there).